Add a new environment machine normalizer

AndrasKovacs opened the pull request:
See this [post on discourse](https://discourse.dhall-lang.org/t/nbe-type-checking-conversion-checking/55) for motivation and background for this PR.

- Dhall.Eval: new evaluator, conversion checker and normalizer. Standalone alpha normalizer is not yet available here.
- There is a new option "new-normalize" for the dhall executable, which uses
  the new normalizer.
- Type checker and the rest of the Dhall codebase remains unchanged.
- There are some benchmarking notes in `dhall/benchmark/examples/new-normalize`. There should be more benchmarks.

AndrasKovacs commented:
Old GHCs trip up on `Strict` pragma. Should I just remove it?

jneira commented:
@AndrasKovacs First of all congrats for this awesome improvement to dhall-haskell
I suppose@Gabriel439 is already aware of it. IMO the final change should be applied directly for default normalization, after the appropiate tests and benchmarks.

Re Strict: I am afraid that remove the pragma would need a massive change in the modules cause you should add explicit strict annotations everywhere (following the semantics of the pragma: https://gitlab.haskell.org/ghc/ghc/wikis/strict-pragma#strict)

AndrasKovacs commented:
@jneira adding strictness annotations is no big deal. On the other hand, using the new normalizer in the codebase is a substantial change, and ideally I'd like to defer that work to a new branch on `dhall-haskell`.

Gabriel439 commented:
@AndrasKovacs: Sorry for the delay; I was a bit sick recently.

The reason CI builds against GHC 7.10.3 is to support Etlas (the package manager for Eta).  Eta/Etlas are built with GHC 7.10.3 and Etlas depends on Dhall (since Dhall is an officially supported configuration format for configuring Etlas packages).  So we have to figure out some way to merge this will supporting GHC 7.10.3.

I can think of two main solutions:

* Configure the `dhall` package to only enable the new module for newer GHC versions
* Disable `Strict` for now if the performance loss isn't too large

My weak preference is for the latter solution, since it seems like the other changes in this pull request are responsible for the large performance gains.

Gabriel439 commented:
@AndrasKovacs: Instead of adding a `new-normalize` subcommand, would it be possible to always the new normalization algorithm for the command-line interpreter?  My understanding is that the only reason we might need to preserve the old normalization code at this point is to support custom normalization, but we don't take advantage of custom normalization when using the command line executable.

I was actually going to suggest going a step further and replacing `Dhall.Core.normalize` with your faster algorithm, but leaving `Dhall.Core.normalizeWith` to use the old customizable normalization algorithm.

AndrasKovacs commented:
Thanks for the feedback!

I'll remove the `Strict` pragma (and conform to GHC 7.10 if there's some other issue, e.g. I think I used newer-style pattern synonyms as well), and I'll look into replacing `Dhall.core.normalize`. 

If that works, it's OK for now, but that's still overall very inefficient with the current typechecker and import resolver. I plan to switch them as well to the standard NbE algorithm, and I think I can do this sometime in the next 2 weeks.

Gabriel439 commented:
@AndrasKovacs: Yeah, I'm fine merging this even with just the improvement to normalization performance.  From my point of view the main reason to merge this is to reduce your maintenance burden since once you merge this into `master` we can ensure that it stays up to date

AndrasKovacs commented:
Updates:

- Compiles with GHC 7.10
- In `Dhall.Core`, `normalize`, `judgmentallyEqual` and `alphaNormalize` have new implementations.
- `normalizeWith` takes `Maybe ReifiedNormalizer` as input (instead of a function), to make it possible to switch between new and old normalizers, depending on whether the custom normalizer parameter is a no-op. 
- Removed `new-normalize` command and the separate tests for the new normalizer which I previously added.
- Added a new standalone `alphaNormalize` implementation, which behaves like the old one, in order to let `TypeCheck` and `Import` function without notable changes for now.

I also tested Vanessa McHale's [big file](https://github.com/vmchale/cpkg/blob/master/pkgs/pkg-set.dhall), and we only have about 25% checking time reduction on that so far. Small computation-heavy examples seem to get far greater speedups. So I believe currently the practically most relevant overheads are in type checking and imports.

jneira commented:
wow appveyor has reached the 1 hour limit build 😟

f-f commented:
@jneira we should probably enable the cache on failed builds too, with `APPVEYOR_SAVE_CACHE_ON_ERROR: true` ([see docs here](https://www.appveyor.com/docs/build-cache/#saving-cache-for-failed-build))

jneira commented:
@f-f maybe it could improve times in case of failed builds but i think the build times has increased since we are using ghc-8.6.4.
I am testing build times [with lts-13 and lts-12](https://ci.appveyor.com/project/jneira/dhall-haskell/builds/23838407), if it is the cause we could use lts-12 for now

Gabriel439 commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+    Merge x y ma     -> case (evalE x, evalE y, evalE <$> ma) of
+                          (VRecordLit m, VUnionLit k v _, _)
+                            | Just f <- Dhall.Map.lookup k m -> f `vApp` v
+                            | otherwise -> error "eval: impossible"
```
Instead of returning `error`, would it be possible to return the fallback case (i.e. `VMerge x y ma`)?  Then the property test in `Dhall.Test.QuickCheck` could be re-enabled

Same comment for other uses of `error`

AndrasKovacs replied:
I prefer not being silent about type errors in evaluation. This is an important point for detecting obscure bugs in type checking. Also, once we compute past an ill-typed expressions, totality and soundness of conversion checking fail to hold.

BTW, I see now that `Import.Types` has an `InternalError` exception. Should I use that instead of these ad-hoc "impossible" errors?  In this case I do intend to signal internal error, and not something like an obviously impossible branch. In Agda though, these are also called "impossible" errors.

Gabriel439 replied:
Yeah, you can use the `Dhall.Core.internalError` function for this purpose.  It provides a standard template for users to report implementation errors

AndrasKovacs replied:
However, I like `error` because it gives me line numbers, while `throw InternalError` doesn't. Is there a way to get line numbers with thrown exceptions?

AndrasKovacs replied:
Or just use `error` with a nicer message...

AndrasKovacs replied:
I used `error` with nicer message in new commit.

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+                                     (NBind e x, Binding "_" (goEnv e <$> a) (goEnv e t):bs))
+                                 (e, []) ts of
+                           (e', Prelude.reverse -> t':ts') -> Let (t' :| ts') (goEnv e' u)
+                           _ -> error "alphaNormalize.goEnv: impossible"
```
You're right, I'll use this version.

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+    (VIntegerToDouble t , VIntegerToDouble t') -> convE t t'
+
+    (VDouble       , VDouble)        -> True
+    (VDoubleLit n  , VDoubleLit n')  -> n == n'
```
OK. I think now I should just export a `doubleToTerm :: Double -> Term` function from `Dhall.Binary`, and use that (I don't want to use `encode` because of the unnecessary `ToTerm` constraint on `conv`; also, the next iteration of `conv` likely will not have any class constraint, because it'll have access to `Val`-s of imported expressions).

Gabriel439 replied:
Another way to do this without a `ToTerm` constraint is to wrap the `Double` value in a `DoubleLit` constructor and then `Dhall.Binary.encode` that

AndrasKovacs replied:
Right, I realize now I just have to specify an arbitrary `Expr s a` type for the `DoubleLit`.

AndrasKovacs replied:
I pushed a commit, I used `encode` as you suggested.

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+import Dhall.Core
+import Dhall.Map (Map)
+import Dhall.Set (Set)
+import GHC.Natural
```
Sure.

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+data Void
+
+coeExprVoid :: Expr Void a -> Expr s a
+coeExprVoid t = unsafeCoerce t
```
Yes, eta-reduction makes no difference here at all.

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+    is fairly cheap to determine at evaluation time.
+
+Potential optimizations with changing Expr:
+  - Use Int in Var instead of Integer. Overflow is practically impossible.
```
OK

AndrasKovacs commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -0,0 +1,900 @@
+    Let (b :| bs) t  -> go env (b:bs) where
+                          go !env []     = eval env t
+                          go  env (b:bs) = go (Extend env (variable b)
+                                                          (eval env (value b))) bs
```
No, we add new entries to the environment as we process the `Let` block, because definitions in the block and the `Let` body may depend on previous definitions. `evalE` is shorthand only for non-`Env` changing `eval`.

AndrasKovacs commented on dhall/dhall.cabal:
```diff
@@ -441,7 +442,8 @@ Library
         Dhall.Tutorial,
-        Dhall.TypeCheck
+        Dhall.TypeCheck,
+        Dhall.Eval
```
OK

ocharles commented:
I was curious to see how this performed with one project that is currently unusable with Dhall's current performance - replacing Nix with Dhall. The issue is at https://github.com/dhallix/nix-derivation/issues/8. I updated the `gcc` branch to `gcc-latest-dhall`, but unfortunately we're still looking at 1m 47s to call `dhall <<< ./gcc.dhall`.

AndrasKovacs commented:
@ocharles: that's a very good benchmark, I'll keep it in mind. I did not expect dramatic speedup from this version, because the typechecker is not yet changed. I'm confident your example will get much faster after I do some work on type checking, which I plan to do relatively soon.

Gabriel439 commented:
Yeah, I know that on Vanessa's example, 80% of the time was spent in the type-checking phase so a 25% increase in performance on that example exactly matches the best case possible improvement without changing type-checking.

ocharles commented:
I did wonder if that were the case. As I know this does type check, I may try dropping type checking out to see how much time is spent normalising.

ocharles commented:
Ok, I think I've disabled the main type checking, and I see that `gcc.dhall` normalisation goes from taking 23.821s to 10.563s. So that's a pretty good improvement! It seems that type checking is the real pain point in my work, so hopefully future work can bring that time down.

ysangkok commented:
@AndrasKovacs the Discourse link in your PR description is down, do you know where the information that was there can be found? Thanks.

AndrasKovacs commented:
@ysangkok the Discourse history was deleted a while ago accidentally. Unfortunately I can't find the page now. Perhaps someone with more web cache/archive expertise can find it.
